Automattic/wp-calypso
# Add 404 universal page for partial matched routes

Fixes #1135. 

Create a new React component (`Page-404`) which renders the "Page not found" text. Template for this page originates in `server/pages/404.jade`. Render this component when a partially correct route is matched.

I'm not sure if this is the right approach to handle the 404 page cases for partially matched routes and would like to start a conversation. For more background, please read the #1135 issue.

## Edit: More info

On WP.com, if you go to an URL that is matches neither any Calypso section or is any other valid URL, WP.com's 404 page is shown (i.e. https://wordpress.com/gibberish). It's because (I was told, don't have access) there is some Nginx whitelist of URLs which simply get passed to Calypso, others to non-Calypso sites and the rest is 404.

However, when the URL is partially correct (e.g. `/stats123`), Nginx redirects it to Calypso, which handles it with its router (`page()` calls). There was no 404 route handler so far in Calypso. This PR creates one.

One approach to detecting whether the request results in 404 (which I use here) is to add a `page( '*' )` route after all the other routes, so that when the requests gets as far as to this last route handler, we have 404 situation. But all the sections are loaded, even though we only display 404 (wasting CPU). However, this is not a big issue if code-splitting is enabled, which, by default, is.

The other approach I was thinking about is to match all the valid URLs even before loading sections. We could simply test for all the sections paths, whether the request's URL (pathname) contains them and if not, display 404. But when code-splitting is enabled, this approach is not much faster than the first one and is additional work, so I'm not sure about going this way.

## Return 404 right from the server

Code in this PR handles 404 requests only on the client. If we want to respond to an invalid request directly from server, we need to write code for it on the server. Is it necessary? I can do it but I'm not sure if I should.

## Testing

1. Checkout `add/404-universal-page-for-partial-matched-routes` branch and start Calypso server.
2. Go to a non-existent route, starting with existing path to a section. (e.g. `/stats123`).
3. You should be presented with a "Page not found" text.

- [x] Needs review if this approach is fine

cc @nb
